noxfile.py

```diff
@@ -0,0 +1,77 @@
+import os
+from pathlib import Path
+
+import nox
+
+LINT_FILES = ("hacking/pr_labeler/label.py", "noxfile.py")
+PINNED = os.environ.get("PINNED", "true").lower() in {"1", "true"}
+nox.options.sessions = ("lint",)
+
+
+def install(session: nox.Session, *args, req: str, **kwargs):
+    if PINNED:
+        kwargs.setdefault("env", {})["PIP_CONSTRAINT"] = f"tests/{req}.txt"
+    session.install("-r", f"tests/{req}.in", *args, **kwargs)
+
+
+@nox.session
+def static(session: nox.Session):
+    """
+    Run static checkers
+    """
+    install(session, req="static")
+    session.run("ruff", *session.posargs, *LINT_FILES)
+
+
+@nox.session
+def formatters(session: nox.Session):
+    """
+    Reformat code
+    """
+    install(session, req="formatters")
+    session.run("isort", *session.posargs, *LINT_FILES)
+    session.run("black", *session.posargs, *LINT_FILES)
+
+
+@nox.session
+def formatters_check(session: nox.Session):
+    """
+    Check code formatting without making changes
+    """
+    install(session, req="formatters")
+    session.run("isort", "--check", *session.posargs, *LINT_FILES)
+    session.run("black", "--check", *session.posargs, *LINT_FILES)
+
+
+@nox.session
+def typing(session: nox.Session):
+    install(session, req="typing")
+    session.run("mypy", *session.posargs, *LINT_FILES)
+
+
+@nox.session
+def lint(session: nox.Session):
+    session.notify("static")
+    session.notify("formatters")
+
+
+requirements_files = list(
+    {path.name.replace(".in", "") for path in Path("tests").glob("*in")}
+    - {"constraints"}
+)
+
+
+@nox.session(name="pip-compile", python=["3.10"])
+@nox.parametrize(["req"], requirements_files, requirements_files)
+def pip_compile(session: nox.Session, req: str):
+    # .pip-tools.toml was introduced in v7
+    session.install("pip-tools>=7")
```

Comment:
1) I like to have spacing for visual readability, in dep specs:
```suggestion
    session.install("pip-tools >= 7")
```
2) Maybe make a separate .in+.txt pair for locking pip-tools itself?